kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/JdbcAuthenticationProviderImpl.scala

```diff
@@ -0,0 +1,235 @@
+    debug(configLog("Database password length", jdbcUserPassword.getOrElse("").length.toString))
+    debug(configLog("Query SQL", authQuerySql.orNull))
+
+    // Check if JDBC parameters valid
+    if (driverClass.isEmpty) {
+      throw new IllegalArgumentException("JDBC driver class is not configured.")
+    }
+
+    if (jdbcUrl.isEmpty) {
+      throw new IllegalArgumentException("JDBC url is not configured")
+    }
+
+    if (jdbcUsername.isEmpty || jdbcUserPassword.isEmpty) {
+      throw new IllegalArgumentException("JDBC username or password is not configured")
+    }
+
+    // Check Query SQL
+    if (authQuerySql.isEmpty) {
+      throw new IllegalArgumentException("Query SQL is not configured")
+    }
+    val querySqlInLowerCase = authQuerySql.get.trim.toLowerCase
+    if (!querySqlInLowerCase.startsWith("select")) { // allow select query sql only
+      throw new IllegalArgumentException("Query SQL must start with \"SELECT\"");
+    }
+    if (!querySqlInLowerCase.contains("where")) {
+      warn("Query SQL does not contains \"WHERE\" keyword");
+    }
+    if (!querySqlInLowerCase.contains("${username}")) {
+      warn("Query SQL does not contains \"${username}\" placeholder");
+    }
+  }
+
+  /**
+   * Extract all placeholders from query and put them into a list.
+   *
+   * @param sql
+   * @return
+   */
+  private def getPlaceholderList(sql: String): List[String] = {
+    SQL_PLACEHOLDER_REGEX.findAllMatchIn(sql)
+      .map(m => m.matched)
+      .toList
+  }
+
+  /**
+   * Replace all placeholders as "?"
+   *
+   * @param sql
+   * @return
+   */
+  private def getPreparedSql(sql: String): String = {
+    SQL_PLACEHOLDER_REGEX.replaceAllIn(sql, "?")
+  }
+
+  /**
+   * prepare the final query statement
+   * by replacing placeholder in query sql with user and password
+   *
+   * @param connection
+   * @param user
+   * @param password
+   * @return
+   */
+  private def getAndPrepareQueryStatement(
+      connection: Connection,
+      user: String,
+      password: String): PreparedStatement = {
+    // Replace placeholders by "?" and prepare the statement
+    val stmt = connection.prepareStatement(getPreparedSql(authQuerySql.get))
+
+    // Extract placeholder list and use its order to pass parameters
+    val placeholderList: List[String] = getPlaceholderList(authQuerySql.get)
+    for (i <- placeholderList.indices) {
+      val param = placeholderList(i) match {
+        case USERNAME_SQL_PLACEHOLDER => user
+        case PASSWORD_SQL_PLACEHOLDER => password
+        case otherPlaceholder =>
+          error(s"Unrecognized Placeholder In Query SQL: $otherPlaceholder")
+          throw new IllegalStateException(
+            s"Unrecognized Placeholder In Query SQL: $otherPlaceholder")
+      }
+
+      stmt.setString(i + 1, param)
+    }
+
+    // Client side limit 1
+    stmt.setMaxRows(1)
+
+    stmt
+  }
+
+  /**
+   * Gracefully close DB connection
+   *
+   * @param connection
+   * @param statement
```

Comment:
please remove the unnecessary dummy scaladoc, it helpless

Reply:
agree.

Reply:
let's do such cleanup on other method as well

Reply:
sure. please have a check in latest commit.